File: src/jimn/tree/inclusion_tree/polygonsegment.py
"""
extend segment class with additional attributes.
"""
from jimn.segment import Segment
from jimn.utils.iterators import all_two_elements


def polygon_segments(height, polygon):
    """
    iterates on all PolygonSegment inside a given polygon at
    a given height
    """
    for start, end in all_two_elements(polygon.points):
        yield PolygonSegment([start, end], height, polygon)


class PolygonSegment(Segment):
    """
    two additional attributes to basic segments : height and polygon we belong to.
    """
    def __init__(self, points, height, polygon_we_belong_to):
        super().__init__(points)
        self.height = height
        self.polygon = polygon_we_belong_to

    def polygon_id(self):
        """
        return id of polygon we belong to.
        """
        return id(self.polygon)

File: rust/jimn/examples/polygon_segment.rs
extern crate jimn;
use jimn::polygon::Polygon;
use jimn::point::Point;
use jimn::segment::Segment;
use jimn::tree::inclusion_tree::polygon_segment::PolygonSegment;

fn main() {
    let p = Polygon::new(vec![
        Point::new(0.0, 0.0),
        Point::new(6.0, 0.0),
        Point::new(3.0, 3.0),
    ]);
    let s = PolygonSegment::new(&p, 0, 0.0);
    println!("{:?}", s);
}
